### PR TITLE
txscript: Implement CheckLockTimeVerify (BIP0065)

### PR DESCRIPTION
**Pull request by dajohi**
This commit mimics Bitcoin Core commit bc60b2b4b401f0adff5b8b9678903ff8feb5867b

### CODE REVIEW

**Comment by dajohi** [comment]
This requires #455 first.

File: txscript/scriptnum.go [hunk 1]
```diff
@@ -8,9 +8,9 @@ const (
 	maxInt32 = 1<<31 - 1
 	minInt32 = -1 << 31
 
-	// maxScriptNumLen is the maximum number of bytes data being interpreted
-	// as an integer may be.
-	maxScriptNumLen = 4
+	// defaultmaxScriptNumLen is the default maximum number of bytes data being
```
Comment (davecgh) [inline txscript/scriptnum.go]: `defaultMaxScriptNumLen`

File: txscript/opcode.go [hunk 2]
```diff
@@ -413,6 +415,8 @@ var opcodeArray = [256]opcode{
 	OP_ENDIF:    {OP_ENDIF, "OP_ENDIF", 1, opcodeEndif},
 	OP_VERIFY:   {OP_VERIFY, "OP_VERIFY", 1, opcodeVerify},
 	OP_RETURN:   {OP_RETURN, "OP_RETURN", 1, opcodeReturn},
+	OP_CHECKLOCKTIMEVERIFY: {OP_CHECKLOCKTIMEVERIFY,
```
Comment (davecgh) [inline txscript/opcode.go]: I know the code is generally kept to 80, but I think for consistency, given every other entry, even those that are longer, are on a single line, this one should be too.

File: txscript/opcode.go [hunk 3]
```diff
@@ -1006,6 +1009,85 @@ func opcodeReturn(op *parsedOpcode, vm *Engine) error {
...
+
+	// Now that we know both locktimes are the same type, only
+	// a simple comparison is needed.
+	if int64(lockTime) > int64(vm.tx.LockTime) {
+		str := "locktime requirement not satisfied -- locktime is " +
+			"greater than the transaction locktime: %d > %d"
+		return fmt.Errorf(str, lockTime, vm.tx.LockTime)
+	}
+
+	// Finally,  the lockTime feature can be disabled and thus
```
Comment (davecgh) [inline txscript/opcode.go]: extra space here

File: txscript/opcode.go [hunk 4]
```diff
@@ -1006,6 +1009,85 @@ func opcodeReturn(op *parsedOpcode, vm *Engine) error {
...
+	// OP_CHECKLOCKTIMEVERIFY bypassed if every TxIn has been
+	// finalized by setting its sequence to the maximum value
+	// (math.MaxUint32).  The transaction would be allowed into
+	// the blockchain making the opcode ineffective.
+	//
+	// Testing if the TxIn is not final is sufficient to
+	// prevent this condition. Alternatively,  we could test all
```
Comment (davecgh) [inline txscript/opcode.go]: extra space here

File: txscript/opcode.go [hunk 5]
```diff
@@ -1006,6 +1008,85 @@ func opcodeReturn(op *parsedOpcode, vm *Engine) error {
...
+	}
+
+	// There are two types of lockTime: lock-by-blockheight
+	// and lock-by-blocktime, distinguished by whether
+	// lockTime < LockTimeThreshold.
+	//
+	// We want to make sure to compare the same type of lockTime,
+	// so fail the script if the lockTimes in both the script and
+	// transaction are of different types.
+	if !((vm.tx.LockTime < LockTimeThreshold && int64(lockTime) < int64(LockTimeThreshold)) ||
```
Comment (davecgh) [inline txscript/opcode.go]: I'd personally prefer a more explicit error message here.  Something like:

```Go
		fmt.Errorf("mismatched locktime types - tx locktime %d, stack "+
			"locktime %d", vm.tx.LockTime, lockTime)
```

File: txscript/opcode.go [hunk 6]
```diff
@@ -1006,6 +1008,85 @@ func opcodeReturn(op *parsedOpcode, vm *Engine) error {
...
+		(vm.tx.LockTime >= LockTimeThreshold && int64(lockTime) >= int64(LockTimeThreshold))) {
+		return fmt.Errorf("locktime requirement not satisfied")
+	}
+
+	// Now that we know both locktimes are the same type, only
+	// a simple comparison is needed.
+	if int64(lockTime) > int64(vm.tx.LockTime) {
+		str := "locktime requirement not satisfied -- locktime is " +
+			"greater than the transaction locktime: %d > %d"
+		return fmt.Errorf(str, lockTime, vm.tx.LockTime)
```
Comment (davecgh) [inline txscript/opcode.go]: I'd personally prefer a more explicit error message here.  Something like:

```Go
		fmt.Errorf("mismatched locktime types - tx locktime %d, stack "+
			"locktime %d", vm.tx.LockTime, lockTime)
```

File: txscript/opcode.go [hunk 7]
```diff
@@ -1006,6 +1008,85 @@ func opcodeReturn(op *parsedOpcode, vm *Engine) error {
...
+	// There are two types of lockTime: lock-by-blockheight
+	// and lock-by-blocktime, distinguished by whether
+	// lockTime < LockTimeThreshold.
+	//
+	// We want to make sure to compare the same type of lockTime,
+	// so fail the script if the lockTimes in both the script and
+	// transaction are of different types.
+	if !((vm.tx.LockTime < LockTimeThreshold && int64(lockTime) < int64(LockTimeThreshold)) ||
+		(vm.tx.LockTime >= LockTimeThreshold && int64(lockTime) >= int64(LockTimeThreshold))) {
+		return fmt.Errorf("locktime requirement not satisfied")
```
Comment (davecgh) [inline txscript/opcode.go]: I'd personally prefer a more explicit error message here.  Something like:

```Go
		fmt.Errorf("mismatched locktime types -- tx locktime %d, stack "+
			"locktime %d", vm.tx.LockTime, lockTime)
```

**Comment by davecgh** [comment]
Aside from the minor things I pointed out, this looks good.  I manually decoded each of the tx tests (both valid and invalid) and verified they are testing the intended things.  I also did a thorough comparison audit since this will ultimately affect consensus.  Finally, the test coverage from the tests is sufficient to cover all branches in the new function:

> github.com\btcsuite\btcd\txscript\opcode.go:1016:       opcodeCheckLockTimeVerify       100.0%

**Comment by jrick** [comment]
Consensus wise it appears correct.  Make a couple nits but they're just my suggestions.  Not too happy with copying Core's comments verbatim.

**Comment by dajohi** [comment]
Comments updated w/ help from @davecgh

**Comment by davecgh** [comment]
OK

**Pull request merged**